src/CentrifugeRouter.sol

```diff
@@ -2,97 +2,197 @@
+
+        _approveMax(asset, vault);
+        IERC7540Vault(vault).requestDeposit(lockedRequest, controller, address(this));
+        emit ExecuteLockedDepositRequest(vault, controller, _initiator);
     }
 
     /// @inheritdoc ICentrifugeRouter
-    function claimDeposit(address vault, address user) external {
-        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(user);
-        IERC7540Vault(vault).deposit(maxDeposit, user, user);
+    function claimDeposit(address vault, address receiver, address controller) external protected {
+        uint256 maxDeposit = IERC7540Vault(vault).maxDeposit(controller);
+        IERC7540Vault(vault).deposit(maxDeposit, receiver, controller);
     }
 
     // --- Redeem ---
     /// @inheritdoc ICentrifugeRouter
-    function requestRedeem(address vault, uint256 amount) external {
-        IERC7540Vault(vault).requestRedeem(amount, msg.sender, msg.sender);
+    function requestRedeem(address vault, uint256 amount, address controller, address owner) external protected {
+        IERC7540Vault(vault).requestRedeem(amount, controller, owner);
+    }
+
+    /// @inheritdoc ICentrifugeRouter
+    function claimRedeem(address vault, address receiver, address controller) external protected {
+        uint256 maxRedeem = IERC7540Vault(vault).maxRedeem(controller);
+        IERC7540Vault(vault).redeem(maxRedeem, receiver, controller);
     }
 
+    // --- ERC20 permits ---
     /// @inheritdoc ICentrifugeRouter
-    function claimRedeem(address vault, address user) external {
-        uint256 maxRedeem = IERC7540Vault(vault).maxRedeem(user);
-        IERC7540Vault(vault).redeem(maxRedeem, user, user);
+    function permit(address asset, address spender, uint256 assets, uint256 deadline, uint8 v, bytes32 r, bytes32 s)
+        external
+        protected
+    {
+        try IERC20Permit(asset).permit(_initiator, spender, assets, deadline, v, r, s) {} catch {}
+    }
+
+    // --- ERC20 wrapping ---
+    function wrap(address wrapper, uint256 amount, address receiver) external protected {
+        address underlying = IERC20Wrapper(wrapper).underlying();
+
+        amount = MathLib.min(amount, IERC20(underlying).balanceOf(_initiator));
+        require(amount != 0, "CentrifugeRouter/zero-balance");
+        SafeTransferLib.safeTransferFrom(underlying, _initiator, address(this), amount);
+
+        _approveMax(underlying, wrapper);
+        require(IERC20Wrapper(wrapper).depositFor(receiver, amount), "CentrifugeRouter/deposit-for-failed");
+    }
+
+    function unwrap(address wrapper, uint256 amount, address receiver) external protected {
+        require(receiver != address(0), "CentrifugeRouter/zero-address");
+        amount = MathLib.min(amount, IERC20(wrapper).balanceOf(address(this)));
+        require(amount != 0, "CentrifugeRouter/zero-balance");
+
+        require(IERC20Wrapper(wrapper).withdrawTo(receiver, amount), "CentrifugeRouter/withdraw-to-failed");
```

Comment:
if the transfer of the wraped tokens to the router  happens in a seperate transaction, what prevents a malicious receiver to actually steal the initial receivers balance

Reply:
on `wrap` we actually transfer the tokens to the router, but on `unwrap` we are not transferring the wrapped tokens to the router...

Reply:
So when you use `depositFor`  the `receiver` you passed is actually used in the account balancing in the wrapper. Basically the `balance[receiver]+=amount`

When you try to `withdrawTo` actually it checks the balance of the `msg.sender`.  So the only way to withdraw funds is if the `receiver` in the `wrap` call actually calls `withdrawTo`.

I think there is another issue thought.

When `wrap` is called it will accrue the amount to the `receiver` and the CentrifugeRouter balance will be 0.
When `unwrap` is called it won't pass because the balance of `msg.sender` in the wrapper ( in our case CentrifugeRouter ) must be >= `amount` but it's 0.

I've been checking these 2 implementations ( based on our impl in morpho-markets )
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC20Wrapper.sol
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol

Reply:
I don't fully follow the concern. I added an unwrap test, it works fine from what I see: https://github.com/centrifuge/liquidity-pools/pull/320/commits/83f118b4948fa129602745bffdbf750ea463ec72